ethexe/ethereum/src/lib.rs

```diff
@@ -79,162 +75,71 @@ pub(crate) type ExeFiller =
-                    _owner: deployer_address,
-                    _mirror: mirror_address,
-                    _wrappedVara: wvara_address,
-                    _middleware: Address::ZERO,
-                    _eraDuration: U256::from(24 * 60 * 60),
-                    _electionDuration: U256::from(2 * 60 * 60),
-                    _validationDelay: U256::from(60),
-                    _aggregatedPublicKey: (AggregatedPublicKey {
-                        x: GearU256::from_big_endian(public_key_x_bytes),
-                        y: GearU256::from_big_endian(public_key_y_bytes),
-                    })
-                    .into(),
-                    _verifiableSecretSharingCommitment: Bytes::copy_from_slice(
-                        &verifiable_secret_sharing_commitment.serialize()?.concat(),
-                    ),
-                    _validators: validators,
-                }
-                .abi_encode(),
-            ),
-        )
-        .await?;
-        let router_address = *proxy.address();
-        let router = IRouter::new(router_address, provider.clone());
-
-        let mirror = IMirror::deploy(provider.clone(), router_address).await?;
-
-        let builder = wrapped_vara.approve(router_address, U256::MAX);
-        builder.send().await?.try_get_receipt().await?;
-
-        assert_eq!(router.mirrorImpl().call().await?, *mirror.address());
-
-        let builder = router.lookupGenesisHash();
-        builder.send().await?.try_get_receipt().await?;
-
-        log::debug!("Router impl has been deployed at {}", router_impl.address());
-        log::debug!("Router proxy has been deployed at {router_address}");
-
-        log::debug!(
-            "WrappedVara impl has been deployed at {}",
-            wrapped_vara_impl.address()
-        );
-        log::debug!("WrappedVara deployed at {wvara_address}");
-
-        log::debug!("Mirror impl has been deployed at {}", mirror.address());
-
+    pub async fn from_provider(provider: AlloyProvider, router: Address) -> Result<Self> {
+        let router_query = RouterQuery::from_provider(router, provider.root().clone());
         Ok(Self {
-            router_address,
-            wvara_address,
+            router,
+            wvara: router_query.wvara_address().await?,
             provider,
         })
     }
+}
 
-    pub fn provider(&self) -> AlloyProvider {
+impl Ethereum {
+    fn provider(&self) -> AlloyProvider {
         self.provider.clone()
     }
 
+    pub fn mirror(&self, address: LocalAddress) -> Mirror {
+        Mirror::new(address.0.into(), self.provider())
+    }
+
     pub fn router(&self) -> Router {
-        Router::new(
-            self.router_address,
-            self.wvara_address,
-            self.provider.clone(),
-        )
+        Router::new(self.router, self.wvara, self.provider())
     }
+}
 
-    pub fn mirror(&self, address: LocalAddress) -> Mirror {
-        Mirror::new(address.0.into(), self.provider.clone())
+impl EthereumWithMiddleware {
```

Comment:
What the profit of using different structure here? Even more I can see disadvantage to make like this. For example in ethexe/service/src/tests/utils/env.rs we have field eth: Ethereum. But in some tests we may wanna to use deployment with middleware - we should we do in that case?
In case Ethereum would contain middleware as option it would be much simplier to operate with that in code
```
struct Ethereum {
...
middleware: Optino<Address>,
}

impl Ethereum {
     pub fn middleware() -> Option<Middleware> {
            ...
     }
}
```